### PR TITLE
fix: `DROP DATABASE` doesn't clean up the source stream job (in v1)

### PR DESCRIPTION
**Pull request by xxchan**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
 `DROP DATABASE` doesn't clean up the source stream job

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by xxchan** [comment]
* **#16825** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/16825?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#18033** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/18033?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @xxchan and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/stream/src/executor/source/source_backfill_executor.rs [hunk 1]
```diff
@@ -641,13 +644,19 @@ impl<S: StateStore> SourceBackfillExecutorInner<S> {
...
+        );
         Ok(states
             .values()
             .all(|state| matches!(state, BackfillState::Finished))
-            && self
-                .backfill_state_store
-                .scan()
-                .await?
+            && !persisted_states.is_empty()
+            && persisted_states
```
Comment (xxchan) [inline src/stream/src/executor/source/source_backfill_executor.rs:659]: 3 parallelism, 2 kafka partitions:

We can see `persisted_states` is from 0 -> 1 -> 2

image

Comment (xxchan) [reply src/stream/src/executor/source/source_backfill_executor.rs:659]: Wait, so checking `!is_empty()` seems not enough. We need to check `len()==num_splits`? 🤡

Comment (xxchan) [reply src/stream/src/executor/source/source_backfill_executor.rs:659]: Or does it goes from 0->1->N? If so, checking `len()>1` might be ok 🤔

Comment (xxchan) [reply src/stream/src/executor/source/source_backfill_executor.rs:659]: It turns out to be the case:
- first time 0
- second time 1
- third time N

image

Comment (xxchan) [reply src/stream/src/executor/source/source_backfill_executor.rs:648]: We will fix this by disallowing online scaling https://github.com/risingwavelabs/risingwave/issues/18300

File: src/meta/src/manager/catalog/mod.rs [hunk 2]
```diff
@@ -625,10 +625,15 @@ impl CatalogManager {
...
                 .await;
 
-            let catalog_deleted_ids = tables_to_drop
+            let streaming_job_deleted_ids = tables_to_drop
                 .into_iter()
                 .filter(|table| valid_table_name(&table.name))
                 .map(|table| StreamingJobId::new(table.id))
+                .chain(
+                    sources_to_drop
+                        .iter()
```
[the remark this answers is not in the document]
Comment (xxchan) [reply src/meta/src/manager/catalog/mod.rs]: done

**Review by yezizp2012 — APPROVED** [review]

**Pull request merged**